src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,173 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public boolean add(T value) {
-        return false;
+        linkNode(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexPosition(index);
+        if (index == size) {
+            linkNode(value);
+            return;
+        }
+        linkNode(value, getNodeByIndex(index));
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementPosition(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementPosition(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldItem = node.item;
+        node.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementPosition(index);
+        return unlinkNode(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        return unlinkNode(object);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void linkNode(T item, Node<T> nearestNode) {
+        Node<T> previousNode = nearestNode.prev;
+        Node<T> newNode = new Node<>(previousNode, item, nearestNode);
+        nearestNode.prev = newNode;
+        if (previousNode != null) {
+            previousNode.next = newNode;
+        } else {
+            first = newNode;
+        }
+        size++;
+    }
+
+    private void linkNode(T item) {
+        Node<T> currentLastNode = last;
+        Node<T> newLastNode = new Node<>(currentLastNode, item, null);
+        last = newLastNode;
+        if (currentLastNode != null) {
+            currentLastNode.next = newLastNode;
+        } else {
+            first = newLastNode;
+        }
+        size++;
+    }
+
+    private T unlinkNode(Node<T> node) {
+        Node<T> previousNode = node.prev;
+        Node<T> nextNode = node.next;
+        if (previousNode != null) {
+            previousNode.next = nextNode;
+            node.prev = null;
+        } else {
+            first = nextNode;
+        }
+        if (nextNode != null) {
+            nextNode.prev = previousNode;
+            node.next = null;
+        } else {
+            last = previousNode;
+        }
+        size--;
+        return node.item;
+    }
+
+    private boolean unlinkNode(T item) {
```

Comment:
do we need this method? better move all logic from this method to the method remove(T object)